aten/src/ATen/native/Sorting.cpp

```diff
@@ -201,6 +202,90 @@ std::tuple<Tensor, Tensor> kthvalue(
   return std::make_tuple(values, indices);
 }
 
+std::tuple<Tensor&, Tensor&> topk_out_cpu(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t k,
+    int64_t dim_,
+    bool largest,
+    bool sorted) {
+  int64_t dim = maybe_wrap_dim(dim_, self.dim(), /*wrap_scalar=*/true);
+  AT_CHECK(
+      k >= 0 && k <= (self.dim() > 0 ? self.size(dim) : 1),
+      "selected index k out of range");
+
+  _allocate_or_resize_output_with_indices(values, indices, self, dim_, k);
+  if (self.dim() == 0 && self.numel() == 1) {
+    values.copy_(self);
+    indices.zero_();
+    return std::forward_as_tuple(values, indices);
+  }
+  auto tmp_values = self.clone();
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "topk_cpu", [&] {
+    dim_apply(
+        {tmp_values, values, indices},
+        dim,
+        [&](int64_t i, TensorList tl) {
+          auto tmp_values = tl[0].accessor<scalar_t, 1>();
+          auto mode_values = tl[1].accessor<scalar_t, 1>();
+          auto mode_indices = tl[2].accessor<int64_t, 1>();
+
+          // we want NaN to be sorted as top for numpy compatibility
+          using elem_t = std::pair<scalar_t, int64_t>;
+          std::function<bool(elem_t&, elem_t&)> cmp =
+            [](elem_t& x, elem_t& y) -> bool {
+              return ((!_isnan<scalar_t>(x.first) && _isnan<scalar_t>(y.first)) || (x.first < y.first));
+            };
+          if (largest) {
+            cmp = [](elem_t& x, elem_t& y) -> bool {
+              return ((_isnan<scalar_t>(x.first) && !_isnan<scalar_t>(y.first)) || (x.first > y.first));
+            };
+          }
+
+          std::vector<elem_t> queue;
+          for (int64_t j = 0; j < tmp_values.size(0); j++) {
+            queue.emplace_back(tmp_values[j], j);
```

Comment:
It's unfortunate that we need to copy the data inside the `TensorAccessor`.

@ezyang do you know how hard it would be to make `TensorAccessor` work with `std::sort` and alike? Would that be something we would want to do at some point?

Reply:
You can just get a raw data pointer if it's contiguous and sort in-place. You're unlikely to get `std::sort` working natively on strided stuff though.

Reply:
Well, if `TensorAccessor` respects the [following requirements](https://en.cppreference.com/w/cpp/algorithm/partial_sort), it would work

> ### Type requirements
> 
> - `RandomIt` must meet the requirements of [ValueSwappable](https://en.cppreference.com/w/cpp/named_req/ValueSwappable) and [LegacyRandomAccessIterator](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator).
> - The type of dereferenced `RandomIt` must meet the requirements of MoveAssignable and MoveConstructible.

My question is more like, how much effort would it be to make this happen, if possible

Reply:
Just like @ezyang said, in case `dim=-1`, (sorting dimension is contiguous), it is possible to create `std::vector` directly on the raw pointer. But if not (non-contiguous), it is not going to work.
I measured the time cost of this copy, it is not that big compared to the sorting. And the current `topk` also has copy, [here](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorMoreMath.cpp#L971), since the sorting needs to swap the value/indice.

Anyway, if `TensorAccessor` and `std::vector` has some sort of in place conversion, it would be more convenient.